ms-common-impl/src/main/java/net/trajano/ms/vertx/beans/JwksProvider.java

```diff
@@ -66,84 +68,78 @@ public void buildJwks() {
-            .keyID(tokenGenerator.newToken())
-            .build();
+        try {
+            RsaJsonWebKey rsaJsonWebKey = RsaJwkGenerator.generateJwk(2048);
+            rsaJsonWebKey.setKeyId(cryptoOps.newToken());
+            rsaJsonWebKey.setAlgorithm(AlgorithmIdentifiers.RSA_USING_SHA512);
+            rsaJsonWebKey.setUse("sig");
+            return rsaJsonWebKey;
+        } catch (JoseException e) {
+            throw new InternalServerErrorException(e);
+        }
     }
 
     /**
      * Gets a single signing key.
      *
-     * @return
+     * @return an RSA web key that supports signing.
      */
-    public RSAKey getASigningKey() {
-
-        final List<JWK> keys = getKeySet().getKeys();
-        return (RSAKey) keys.get(random.nextInt(keys.size()));
-    }
-
-    public RSAKey getDecryptionKey(final String keyID) {
-
-        return (RSAKey) getKeySet().getKeyByKeyId(keyID);
+    public RsaJsonWebKey getASigningKey() {
 
+        final JsonWebKeySet keySet = getKeySet();
+        if (LOG.isDebugEnabled()) {
+            LOG.debug(keySet.toJson());
+        }
+        return (RsaJsonWebKey) keySet.findJsonWebKey(null, "RSA", "sig", null);
     }
 
     /**
      * Builds the keys from the cache.
      *
      * @return
      */
-    public JWKSet getKeySet() {
+    public JsonWebKeySet getKeySet() {
 
-        final List<JWK> keys = new LinkedList<>();
+        final JsonWebKeySet set = new JsonWebKeySet();
         for (int i = 0; i < MAX_NUMBER_OF_KEYS; ++i) {
             final String cacheKey = String.valueOf(i);
-            final String jwkJson = jwksCache.get(cacheKey, String.class);
-            try {
-                if (jwkJson != null) {
-                    keys.add(JWK.parse(jwkJson));
-                }
-            } catch (final ParseException e) {
-                LOG.error("unable to parse key={} json={} removing entry", cacheKey, jwkJson);
-                jwksCache.evict(cacheKey);
+            final JsonWebKey jwk = jwksCache.get(cacheKey, JsonWebKey.class);
+            if (jwk != null) {
+                set.addJsonWebKey(jwk);
             }
         }
-        return new JWKSet(keys);
+        return set;
     }
 
     @PostConstruct
     public void init() {
 
+        if (cm == null) {
+            LOG.warn("A org.springframework.cache.CacheManager was not provided an in-memory cache will be used");
+            cm = new ConcurrentMapCacheManager(JWKS_CACHE);
+        }
+
+        jwksCache = cm.getCache(JWKS_CACHE);
         if (jwksCache == null) {
-            LOG.warn("A org.springframework.cache.Cache named {} was not provided an in-memory cache will be used", JWKS_CACHE);
+            LOG.warn("A no cache named {} was not provided by the cache manager an in-memory cache will be used", JWKS_CACHE);
             final ConcurrentMapCacheManager cm = new ConcurrentMapCacheManager(JWKS_CACHE);
             jwksCache = cm.getCache(JWKS_CACHE);
         }
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Rename "cm" which hides the field declared at line 44. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AHiddenFieldCheck)